text/0000-proc-macro-attribute-recursion.md

```diff
@@ -0,0 +1,80 @@
+- Feature Name: proc_macro_attribute_recursion
+- Start Date: 24.01.2019
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Expand `proc_macro_attributes` recursively.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, procedural macros have no way to expand macros at all. [RFC #2320](https://github.com/rust-lang/rfcs/pull/2320) aims to rectify this, but despite being reworked a lot still suffers from some complexity.
+
+This proc_macro author wants something workable now instead of waiting for that RFC while leaving the doors open for an eventual implementation. Also making this small part available allows us to collect experience with macro expansion in proc_macros at very modest cost.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`proc_macro_attributes` can add other attributes in their output. The expander then expands all macros in the proc_macro output, and then checks again for any attributes, expanding them, too (possibly recursively):
+
+Here's an example from [flamer](https://crates.io/crates/flamer):
+
+```rust
+use flamer::flame;
+
+macro_rules! macro_fun {
+    () => {
+        fn this_is_fun(x: u64) -> u64 {
+            x + 1
+        }
+    }
+}
+
+#[flame]
+mod fun {
+    macro_fun!();
+}
+```
+
+Flamer expands the `macro_fun!();` to `#[flame] macro_fun!();`. The expander then expands `macro_fun!()` yielding
+
+```rust
+mod fun {
+    #[flame]
+    fn this_is_fun(x: u64) -> u64 {
+        x + 1
+    }
+}
+```
+
+Because of the `#[flame]` attribute, this is fed back to flamer, which modifies the function resulting from the macro. Note that as in this example, the attribute needs not be placed at the same AST node (and in fact, flamer would place it only on macro invocation nodes).
+
+`proc_macro` writers can implement their macros in terms of `proc_macro_attributes` (which is a very roundabout way to deal with macros, but at least it would work at all) to gain the same benefits.
+
+This way, `proc_macro_attribute`s can be deemed *recursive* like macros-by-example. Note that the macro recursion limit must also be observed by the `proc_macro_attribute` implementations.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The expander is extended to search the expansion of `proc_macro` and `proc_macro_attributes` for other macro invocations. Those are then expanded until there are no more attributes or macro invocations left or the macro expansion limit is reached, whichever comes first.
```

Comment:
This whole section (and the RFC in general) seems rather underspecified; I'd like to see examples of `proc_macro` added to the previous section at least. This also doesn't seem like a full description of behavior since `#[flame]` gets applied to macro expansions. This also doesn't say what happens if you write `#[flame(alpha, beta)]`. This should be specified including with examples.

Reply:
I didn't detail what `#[flame(alpha, beta)]` does (currently it should give you an error), because this RFC does not change this part of functionality. As for bang-macros, it doesn't matter if they are defined via `macro`, `macro_rules!` or as a proc macro. Again the functionality of expansion is unchanged.

The only two things this RFC defines: That the output of `proc_macro_attribute`s get expanded and that the order of expansion follows the order of appearance (so that things from the original code get expanded before things added by the expansion).

Reply:
Well, from what I can tell, the reference says nothing about how `#[flame]` gets attached to `this_is_fun`.

> That the output of proc_macro_attributes get expanded and that the order of expansion follows the order of appearance (so that things from the original code get expanded before things added by the expansion).

That's not what the text says; it says "The expander is extended to search the expansion of `proc_macro` and `proc_macro_attributes` for other macro invocations." -- you've added examples for the latter but not the former.